libvast/builtins/operators/summarize.cpp

```diff
@@ -321,212 +209,368 @@ struct group_by_key_equal {
+  auto add(const table_slice& slice, const configuration& config)
+    -> caf::error {
+    // Step 1: Resolve extractor names (if possible).
+    auto it = bindings.find(slice.schema());
+    if (it == bindings.end()) {
+      auto bound = binding::make(slice.schema(), config);
+      if (!bound) {
+        return bound.error();
+      }
+      it = bindings.try_emplace(it, slice.schema(), std::move(*bound));
+    }
+    auto const& bound = it->second;
+    // Step 2: Collect the aggregation columns and group-by columns into arrays.
+    auto batch = to_record_batch(slice);
+    auto group_by_arrays = bound.make_group_by_arrays(*batch, config);
+    auto aggregation_arrays = bound.make_aggregation_arrays(*batch);
+    // A key view used to determine the bucket for a single row.
     auto reusable_key_view = group_by_key_view{};
-    reusable_key_view.resize(group_by_columns.size(), {});
-    // Helper lambda for finding a bucket, or creating a new one lazily.
-    const auto find_or_create_bucket = [&](int row) -> bucket* {
-      for (size_t column = 0; column < group_by_columns.size(); ++column)
-        reusable_key_view[column] = value_at(group_by_columns[column].type,
-                                             *group_by_arrays[column], row);
-      if (auto bucket = buckets.find(reusable_key_view);
-          bucket != buckets.end())
-        return bucket.value().get();
+    reusable_key_view.resize(bound.group_by_columns.size(), {});
+    // Returns the group that the given row belongs to, creating new groups
+    // whenever necessary.
+    auto find_or_create_bucket = [&](int64_t row) -> caf::expected<bucket*> {
+      for (size_t col = 0; col < bound.group_by_columns.size(); ++col) {
+        if (bound.group_by_columns[col]) {
+          VAST_ASSERT(group_by_arrays[col].has_value());
+          reusable_key_view[col] = value_at(bound.group_by_columns[col]->type,
+                                            **group_by_arrays[col], row);
+        } else {
+          VAST_ASSERT(!group_by_arrays[col].has_value());
+          reusable_key_view[col] = caf::none;
+        }
+      }
+      if (auto it = buckets.find(reusable_key_view); it != buckets.end()) {
+        auto&& bucket = *it->second;
+        // Check that the group-by values also have matching types.
+        VAST_ASSERT(bucket.group_by_types.size()
+                    == bound.group_by_columns.size());
+        for (auto&& [existing, other] :
+             detail::zip{bucket.group_by_types, bound.group_by_columns}) {
+          if (!other) {
+            // If this group-by column does not exist in the input schema, we
+            // assume `null` values and ignore its type.
+            continue;
+          }
+          if (other->type != existing) {
+            // If the types mismatch, we have found equal values with not-equal
+            // types. This can happen e.g. for `null` values. TODO: Prune type?
```

Comment:
We definitely want to prune the type here.

Reply:
Should the corresponding output type then be the pruned type? And under the assumption that my understanding is correct, is it then okay that `summarize ... by src_port` changes `src_port: port` to `src_port: uint64`?

Reply:
Or should we perhaps only prune in the case of a mismatch?

Reply:
Ah, I think I wrote my comment at the same time as you wrote your second one. Yeah, I think pruning on mismatch is best.

Reply:
I think 4bafcf4167671637f901704a41fe874d4b2e761b should do the right thing.

It's not easy to test right now, but after tenzir/issues#505 we can write some nice tests for this.